activerecord/lib/active_record/connection_adapters/query_intent.rb

```diff
@@ -93,29 +161,95 @@ def has_binds?
       end
 
       def execute!
-        adapter.execute_intent(self) # sets our raw_result
+        if can_run_async?
+          async_schedule!(ActiveRecord::Base.asynchronous_queries_session)
+        else
+          @ran_async = false
+          run_query!
+        end
+      ensure
         @executed = true
-        nil
+      end
+
+      def future_result
+        if pending? || can_run_async?
+          FutureResult.new(self)
+        else
+          FutureResult.wrap(cast_result)
+        end
       end
 
       def finish
         affected_rows # just to consume/close the result
         nil
       end
 
+      # Internal setter for raw result
+      def raw_result=(value)
+        @raw_result = value
+        @raw_result_available = true
+      end
+
+      # Check if result has been populated yet (without blocking)
+      def raw_result_available?
+        @raw_result_available
+      end
+
+      # Access the raw result, ensuring it's available first
+      def raw_result
+        ensure_result
+        @raw_result
+      end
+
+      # Ensure the result is available, blocking if necessary
+      def ensure_result
+        if @session
+          # Async was scheduled: wait for result (sets lock_wait)
+          execute_or_wait
+        end
+
+        @event_buffer&.flush
+
+        # Raise any error captured during deferred execution
+        raise @error if @error
+      end
+
       def cast_result
         raise "Cannot call cast_result before query has executed" unless @executed
         raise "Cannot call cast_result after affected_rows has been called" if defined?(@affected_rows)
-        @cast_result ||= adapter.send(:cast_result, raw_result)
+
+        ensure_result
+        @cast_result ||= adapter.send(:cast_result, @raw_result)
```

Comment:
Not new and somewhat tangential, but I think we should make `cast_result` public but `:nodoc:` to avoid that `send`.